src/primitives/gf32.rs

```diff
@@ -0,0 +1,481 @@
+    's', '3', 'j', 'n', '5', '4', 'k', 'h', // +16
+    'c', 'e', '6', 'm', 'u', 'a', '7', 'l', // +24
+];
+
+/// Mapping from bech32 character (either case) to numeric value.
+#[rustfmt::skip]
+const CHARS_INV: [i8; 128] = [
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    15, -1, 10, 17, 21, 20, 26, 30,  7,  5, -1, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+];
+
+/// An element of GF32.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Fe32(pub(crate) u8); // pub(crate) because of `hrpstring` module.
+
+impl Fe32 {
+    // These are a little gratuitous for a reference implementation, but it makes me happy to do it.
+    /// Numeric value maps to bech32 character: 0 == "q".
+    pub const Q: Fe32 = Fe32(0);
+    /// Numeric value maps to bech32 character: 1 == "p".
+    pub const P: Fe32 = Fe32(1);
+    /// Numeric value maps to bech32 character: 2 == "z".
+    pub const Z: Fe32 = Fe32(2);
+    /// Numeric value maps to bech32 character: 3 == "r".
+    pub const R: Fe32 = Fe32(3);
+    /// Numeric value maps to bech32 character: 4 == "y".
+    pub const Y: Fe32 = Fe32(4);
+    /// Numeric value maps to bech32 character: 5 == "9".
+    pub const _9: Fe32 = Fe32(5);
+    /// Numeric value maps to bech32 character: 6 == "x".
+    pub const X: Fe32 = Fe32(6);
+    /// Numeric value maps to bech32 character: 7 == "8".
+    pub const _8: Fe32 = Fe32(7);
+    /// Numeric value maps to bech32 character: 8 == "g".
+    pub const G: Fe32 = Fe32(8);
+    /// Numeric value maps to bech32 character: 9 == "f".
+    pub const F: Fe32 = Fe32(9);
+    /// Numeric value maps to bech32 character: 10 == "2".
+    pub const _2: Fe32 = Fe32(10);
+    /// Numeric value maps to bech32 character: 11 == "t".
+    pub const T: Fe32 = Fe32(11);
+    /// Numeric value maps to bech32 character: 12 == "v".
+    pub const V: Fe32 = Fe32(12);
+    /// Numeric value maps to bech32 character: 13 == "d".
+    pub const D: Fe32 = Fe32(13);
+    /// Numeric value maps to bech32 character: 14 == "w".
+    pub const W: Fe32 = Fe32(14);
+    /// Numeric value maps to bech32 character: 15 == "0".
+    pub const _0: Fe32 = Fe32(15);
+    /// Numeric value maps to bech32 character: 16 == "s".
+    pub const S: Fe32 = Fe32(16);
+    /// Numeric value maps to bech32 character: 17 == "3".
+    pub const _3: Fe32 = Fe32(17);
+    /// Numeric value maps to bech32 character: 18 == "j".
+    pub const J: Fe32 = Fe32(18);
+    /// Numeric value maps to bech32 character: 19 == "n".
+    pub const N: Fe32 = Fe32(19);
+    /// Numeric value maps to bec32 character: 20 == "5".
```

Comment:
In 310012dd3e5f5725b5bde9021606ed1d5542620d:

lol multiple of these misspell `bech32`

Reply:
ha!